src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp

```diff
@@ -44,105 +44,146 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
-        upperBounds(i) = dist;
-        assignments[i] = c;
+        localNewCentroids.col(assignments[i]) += dataset.col(i);
+        ++localCounts(assignments[i]);
+        continue;
       }
-      else if (dist < lowerBounds(i))
+
+      // The bounds failed. So test against all other clusters.
+      lowerBounds(i) = DBL_MAX;
+      size_t newAssignment = assignments[i];
+      double newUpperBound = upperBounds(i);
+      double newLowerBound = DBL_MAX;
+
+      for (size_t c = 0; c < numClusters; ++c)
       {
-        // This is a closer second-closest cluster.
-        lowerBounds(i) = dist;
+        if (c == assignments[i])
+          continue;
+
+        const double dist = distance.Evaluate(dataset.col(i), centroids.col(c));
+
+        // Is this a better cluster?
+        if (dist < newUpperBound)
+        {
+          newLowerBound = newUpperBound;
+          newUpperBound = dist;
+          newAssignment = c;
+        }
+        else if (dist < newLowerBound)
+        {
+          newLowerBound = dist;
+        }
       }
+      localDistanceCalculations += numClusters - 1;
+
+      // Update bounds and assignment
+      upperBounds(i) = newUpperBound;
+      lowerBounds(i) = newLowerBound;
+      assignments[i] = newAssignment;
+
+      // Update new centroids.
+      localNewCentroids.col(newAssignment) += dataset.col(i);
+      ++localCounts(newAssignment);
     }
-    distanceCalculations += centroids.n_cols - 1;
 
-    // Update new centroids.
-    newCentroids.col(assignments[i]) += dataset.col(i);
-    ++counts(assignments[i]);
+    #pragma omp critical
+    {
+      newCentroids += localNewCentroids;
+      counts += localCounts;
+      hamerlyPruned += localHamerlyPruned;
+      distanceCalculations += localDistanceCalculations;
+    }
   }
 
-  // Normalize centroids and calculate cluster movement (contains parts of
-  // Move-Centers() and Update-Bounds()).
+  // Normalize centroids and calculate cluster movement
   double furthestMovement = 0.0;
   double secondFurthestMovement = 0.0;
   size_t furthestMovingCluster = 0;
-  arma::vec centroidMovements(centroids.n_cols);
+  arma::vec centroidMovements(numClusters);
   double centroidMovement = 0.0;
-  for (size_t c = 0; c < centroids.n_cols; ++c)
+
+  #pragma omp parallel for reduction(+:centroidMovement) \
+      reduction(max:furthestMovement, secondFurthestMovement) schedule(static)
+  for (size_t c = 0; c < numClusters; ++c)
   {
     if (counts(c) > 0)
       newCentroids.col(c) /= counts(c);
+    else
+      newCentroids.col(c) = centroids.col(c);
 
     // Calculate movement.
-    const double movement = distance.Evaluate(centroids.col(c),
-                                              newCentroids.col(c));
+    const double movement = std::sqrt(arma::sum(arma::square(centroids.col(c) - newCentroids.col(c))));
     centroidMovements(c) = movement;
     centroidMovement += std::pow(movement, 2.0);
-    ++distanceCalculations;
```

Comment:
This is there for a reason, maybe if you removed it because of a threading issue it should be a `std::atomic`?